daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala

```diff
@@ -26,16 +26,24 @@ import scala.util.{Failure, Success, Try}
 /** Speedy scenario runner that uses the reference ledger.
   *
   * @constructor Creates a runner using an instance of [[Speedy.Machine]].
+  * @param machine the machine use to run the scenario
+  * @param compiledPackages the [[CompiledPackages]] with which the machine was built. It needs to be immutable
+  *                         as the [[ledger]] field is build assume the package never change.
+  * @para initialSeed: initialize the internal [[crypto.Hash]] pseudo random generator.
   */
 final class ScenarioRunner private (
     machine: Speedy.Machine,
+    compiledPackages: PureCompiledPackages,
     initialSeed: crypto.Hash,
 ) {
   import ScenarioRunner._
 
   private[this] val nextSeed: () => crypto.Hash = crypto.Hash.secureRandom(initialSeed)
 
-  var ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
+  private var _ledger: ScenarioLedger =
+    ScenarioLedger(compiledPackages.packageSignature, Time.Timestamp.Epoch)
+  def ledger: ScenarioLedger = _ledger
```

Comment:
what do we gain from splitting `ledger` into `ledger` and `_ledger`?

Reply:
We make sure, that no code outside the object modifies the variable.

Reply:
ah I see it now, thanks!